File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Codec {
private: 
    void serializeHelper(TreeNode *root, stringstream &ss) {
        if (root == nullptr) {
            ss << "#,";
            return;
        }

        ss << root->val << ",";

        serializeHelper(root->left, ss);
        serializeHelper(root->right, ss);

    }

    TreeNode* deserializeHelper(queue<string> &nodes) {
        if (nodes.empty()) {
            return nullptr;
        }

        string node = nodes.front();
        nodes.pop();

        if (node == "#") {
            return nullptr;
        }

        auto root = new TreeNode(stoi(node));
        root->left = deserializeHelper(nodes);
        root->right = deserializeHelper(nodes);
        return root;
    }

public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        stringstream ss;
        serializeHelper(root, ss);

        string result = ss.str();

        result.pop_back();
        return result;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        stringstream ss(data);
        queue<string> nodes;
        string node;

        while (getline(ss, node, ',')) {
            nodes.push(node);
        }

        return deserializeHelper(nodes);
    }
};

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    def serialize(self, root: Optional[TreeNode]) -> str:

        result = []

        def dfs(node):
            if node is None:
                result.append('N')
                return
            
            result.append(str(node.val))
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return ",".join(result)

    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        vals = data.split(",")
        self.idx = 0

        def dfs():
            if vals[self.idx] == 'N':
                self.idx += 1
                return None
            
            node = TreeNode(val=int(vals[self.idx]))
            self.idx += 1
            node.left = dfs()
            node.right = dfs()

            return node

        
        root = dfs()
        return root

        



